### PR TITLE
Update cythonization in CI

### PR DESCRIPTION
**Pull request by jrbourbeau**
This is just a sanity check to see what CI test builds are using the cythonized scheduler. cc @jakirkham @crusaderky 

xref https://github.com/dask/distributed/issues/4760#issuecomment-828519228

### CODE REVIEW

**Comment by jrbourbeau** [comment]
Note that I expect this to fail for all builds except Python 3.7 as that's the one build in which we are cythonizing things

https://github.com/dask/distributed/blob/ea5cf0ed629c3969e5eaf0af2d173040f94a1113/.github/workflows/tests.yaml#L51-L54

**Comment by jrbourbeau** [comment]
Alright, as @jakirkham mentioned in https://github.com/dask/distributed/issues/4760#issuecomment-828602691, our current CI setup isn't testing the cythonized scheduler correctly. I've updated our CI installation step and added a small test to ensure that we're cythonizing the scheduler as expected.

**Comment by jakirkham** [comment]
Yeah that doesn't surprise me. Had tried to get this to work back when GitHub Actions was only used for Windows ( https://github.com/dask/distributed/pull/4326 ), but ran into difficulties. Not having access to Windows locally and the fact we are primarily Cythonizing the Scheduler on Linux, made it hard to prioritize further work to fix CI. Agree it would be good to get this working and it may make more sense to do now as we are making fewer changes to the Scheduler

**Comment by jrbourbeau** [comment]
Sounds good, thanks for taking a look at this @jakirkham. I'm waiting for `test_cythonized` to pass on all CI builds and then I'll run the full test suite to see what happens

**Comment by jakirkham** [comment]
I think one of the issues that we are likely running into here is the fact that we have both a `.py` and a `.so` file, both of which have the same name. So how Python determines, which one is loaded affects whether the Cythonized Scheduler is used. On Linux and macOS, it typically seems to load `.so` files preferentially. On Windows this seems to be the other way around. Ideally we wouldn't leave it up to the interpreter to choose between these two. Part of what that PR was trying to do was trying to do was ensure only one of these files is installed

**Comment by jrbourbeau** [comment]
Ah, sorry I missed that you linked to https://github.com/dask/distributed/pull/4326. Would you prefer for us to move back to that PR instead of what I have here?

**Comment by jakirkham** [comment]
Not at all. That was more for context. The linked PR is pretty old and likely needs to be refreshed. Happy to just start here with what you have 🙂

**Comment by jakirkham** [comment]
Woot! 🎉 Looks like we are now reproducing issue on CI ( https://github.com/dask/distributed/issues/4760 )

```
Got error on 'http://localhost:8787/status/': HTTPError: 500 Server Error: Internal Server Error for url: http://localhost:8787/status/
...
distributed/cli/tests/test_dask_scheduler.py::test_dashboard
```

**Comment by jrbourbeau** [comment]
Okay, so it looks like the `distributed/cli/tests/test_dask_scheduler.py::test_dashboard` failures for the cythonized builds are related to https://github.com/dask/distributed/issues/4760 and I can confirm that locally when I check out https://github.com/dask/distributed/pull/4761 and use the cythonized scheduler, `distributed/cli/tests/test_dask_scheduler.py::test_dashboard` passes

**Comment by jakirkham** [comment]
So how should we handle incorporating these PRs? Does GitHub Actions have a concept of allowed failures? Or are we ok with having a failure that we clear out with a later PR?

**Comment by jrbourbeau** [comment]
Hmm good question. I'd like to merge https://github.com/dask/distributed/pull/4761 and then merge `main` here to see if there are other failures. If there are, we can open up separate PRs to fix them, merge them into `main`, and then check back here until everything is green. What do you think about that?

**Comment by jakirkham** [comment]
SGTM. Happy to dig into any additional failures we find.

FWIW have usually been running the Scheduler tests locally and sometimes other components (like Client and Worker tests), but not always the full test suite. So it is possible that we encounter some failures that are a bit more off this beaten path.

**Comment by jakirkham** [comment]
Had some lingering changes to add a global variable to the Scheduler file indicating whether it was compiled or not. Cleaned them up and submitted them as PR ( https://github.com/jrbourbeau/distributed/pull/2 ) based off this one. This should make it a bit easier to check whether the Scheduler was Cythonized. Also this can be used to mark tests as `xfail`ed when the Scheduler is Cythonized if needed as well. Hopefully that will let us get this change in sooner without needing to wait for individual tests failures to go in beforehand

File: distributed/tests/test_scheduler.py
Comment (jakirkham) [inline distributed/tests/test_scheduler.py:75]: Sorry missed dropping the `import` in that PR

```suggestion
    # This checks whether or not CI builds (where we set the CYTHONIZED environment variable) are
    # cythonized as expected
    assert strtobool(os.environ["CYTHONIZED"]) == COMPILED
```

Comment (jrbourbeau) [reply distributed/tests/test_scheduler.py]: No worries -- we should be good now 👍

**Comment by jrbourbeau** [comment]
It looks like some tests are failing because `distributed.is_coroutine_function` isn't recognizing some `Scheduler` methods, e.g. `Scheduler.restart`, as coroutine functions when the scheduler is cythonized.

With cythonization:

```python
In [1]: from distributed import Scheduler

In [2]: Scheduler.restart
Out[2]: <cyfunction Scheduler.restart at 0x120db4d40>

In [3]: from distributed.utils import is_coroutine_function

In [4]: is_coroutine_function(Scheduler.restart)
Out[4]: False
```

Without cythonization:

```python
In [1]: from distributed import Scheduler

In [2]: Scheduler.restart
Out[2]: <function distributed.scheduler.Scheduler.restart(self, client=None, timeout=3)>

In [3]: from distributed.utils import is_coroutine_function

In [4]: is_coroutine_function(Scheduler.restart)
Out[4]: True
```

Is there a way we can have `cython` maintain coroutine functions?

**Comment by jrbourbeau** [comment]
Another class of failures is due to a `cloudpickle` serialization issue:

```python

Traceback (most recent call last):
  File "/home/runner/work/distributed/distributed/distributed/protocol/serialize.py", line 329, in serialize
    header, frames = dumps(x, context=context) if wants_context else dumps(x)
  File "/home/runner/work/distributed/distributed/distributed/protocol/serialize.py", line 55, in pickle_dumps
    protocol=context.get("pickle-protocol", None) if context else None,
  File "/home/runner/work/distributed/distributed/distributed/protocol/pickle.py", line 60, in dumps
    result = cloudpickle.dumps(x, **dump_kwargs)
  File "/usr/share/miniconda3/envs/dask-distributed/lib/python3.7/site-packages/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/usr/share/miniconda3/envs/dask-distributed/lib/python3.7/site-packages/cloudpickle/cloudpickle_fast.py", line 563, in dump
    return Pickler.dump(self, obj)
_pickle.PicklingError: Can't pickle <cyfunction Scheduler.__init__.<locals>.<lambda> at 0x7f7cd474d390>: attribute lookup lambda5 on distributed.scheduler failed
```

**Comment by jakirkham** [comment]
Thanks James! 😄 WDYT about marking all of these xfail and merging this PR?

I think the issue in these cases is Cython constructing functions for some of these more atypical cases (closures, lambdas, coroutines, etc.) such that they work in C, but that may come with some tradeoffs (like not being able to pickle them). Some of these might be solved with a small rewrite or similar. Some of these may be bugs in Cython. Will take a look and see what we can do for these.

**Comment by jrbourbeau** [comment]
> WDYT about marking all of these xfail and merging this PR?

+1 marking tests right now

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks James! 😄

**Comment by jakirkham** [comment]
Opened PR ( https://github.com/dask/distributed/pull/4767 ) to start working on fixing the skipped/xfailed tests

**Comment by jakirkham** [comment]
Fixing a couple of pickling issues with PR ( https://github.com/dask/distributed/pull/4768 )

**Comment by jakirkham** [comment]
> It looks like some tests are failing because `distributed.is_coroutine_function` isn't recognizing some `Scheduler` methods, e.g. `Scheduler.restart`, as coroutine functions when the scheduler is cythonized.
> 
> With cythonization:
> 
> ```python
> In [1]: from distributed import Scheduler
> 
> In [2]: Scheduler.restart
> Out[2]: <cyfunction Scheduler.restart at 0x120db4d40>
> 
> In [3]: from distributed.utils import is_coroutine_function
> 
> In [4]: is_coroutine_function(Scheduler.restart)
> Out[4]: False
> ```
> 
> Without cythonization:
> 
> ```python
> In [1]: from distributed import Scheduler
> 
> In [2]: Scheduler.restart
> Out[2]: <function distributed.scheduler.Scheduler.restart(self, client=None, timeout=3)>
> 
> In [3]: from distributed.utils import is_coroutine_function
> 
> In [4]: is_coroutine_function(Scheduler.restart)
> Out[4]: True
> ```
> 
> Is there a way we can have `cython` maintain coroutine functions?

Not sure atm, but did find an MRE. Filed as issue ( https://github.com/cython/cython/issues/4138 )

**Comment by da-woods** [comment]
I've posted this on the Cython bug but: `asyncio.iscoroutinefunction` is more flexible that `inspect.iscoroutinefunction` and recent versions of Cython will pass the test. 

https://github.com/dask/distributed/blob/56aed4489288fc95f36e9fe11f29a452c39de4ac/distributed/utils.py#L1264

I think the consensus for Cython is that the `inspect` test is too restrictive, and Cython can't pass that.

**Comment by jakirkham** [comment]
Yeah that makes sense. Thanks for the advice 🙂

Was looking through the `inspect` module, but wasn’t seeing a good way to extend it.

Think we could use `asyncio` here. We already have a utility function for this kind of check where we could make that change pretty easily

**Comment by jakirkham** [comment]
Submitted PR ( https://github.com/dask/distributed/pull/4771 ) to switch to the `asyncio` implementation. Thanks again for the suggestion 🙂

**Pull request merged**
